File: leetcode/面试题04.06.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

/*
 * 面试题 04.06
 */
func inorderSuccessor(root *TreeNode, p *TreeNode) *TreeNode {
	if root == nil {
		return nil
	}
	var res *TreeNode
	cur := root
	for cur != nil {
		if cur.Val > p.Val {
			res = cur
			cur = cur.Left
		} else {
			cur = cur.Right
		}
	}
	return res
}

File: leetcode/面试题04.06.py
#
# 面试题 04.06
#
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def inorderSuccessor(self, root: TreeNode, p: TreeNode) -> TreeNode:
        if not root:
            return
        cur, res = root, None
        while cur:
            if cur.val > p.val:
                res = cur
                cur = cur.left
            else:
                cur = cur.right
        return res
